File: tenka1-2014-quala/b.py
class Event(object):
    def __init__(self, back=0, act=True, t=0, hit=0):
        self.back = back
        self.act = act
        self.throw= t
        self.hit = hit

def damage(v):
    return (int)(v * (1 + combo // 10 * 0.1))

def N(i):
    global kabu, events
    if not kabu:
        return
    kabu -= 1
    events[i + 7].back = 1
    events[i + 1].throw = 10
    events[i + 2].hit = 1

def C(i):
    global kabu, events
    if kabu < 3:
        return
    kabu -= 3
    events[i + 9].back = 3
    events[i + 3].throw = 50
    events[i + 4].hit = 1
    events[i + 1].act = events[i + 2].act = False

S = input()
n = len(S)
events = [Event() for i in range(n + 10)]
combo = 0
kabu = 5
d = 0
for i in range(n + 10):
    kabu += events[i].back
    d += damage(events[i].throw)
    combo += events[i].hit

    if i >= n or not events[i].act or S[i] == '-':
        continue
    if S[i] == 'N':
        N(i)
    if S[i] == 'C':
        C(i)
print(d)

File: tenka1-2014-quala/b.cpp
#include<queue>
#include<string>
#include<utility>
#include<vector>
#include<functional>
#include<iostream>
using namespace std;

int kabu = 5, combo, damage, n;
string s;
bool stop[1001000];
priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> th, dm, back;
int main()
{
  cin >> s;
  n = s.length();
  for(int i = 0; i < n + 10; i++)
  {
    while(!th.empty() && th.top().first == i)
    {
      dm.push(make_pair(
            th.top().first + 1, 
            (int)(th.top().second * (1 + (combo / 10) * 0.1))));
      th.pop();
    } 
    while(!dm.empty() && dm.top().first == i)  
    {
      damage += dm.top().second;
      dm.pop();
      combo++;
    }
    while(!back.empty() && back.top().first == i)
    {
      kabu += back.top().second;
      back.pop();
    }
    
    if(i >= n || stop[i] || s[i] == '-')
      continue;
    switch(s[i])
    {
      case 'N':
        if(!kabu)
          break;
        th.push(make_pair(i + 1, 10));
        back.push(make_pair(i + 7, 1));
        kabu--;
        break;
      case 'C':
        if(kabu < 3)
          break;
        th.push(make_pair(i + 3, 50)); 
        back.push(make_pair(i + 9, 3));
        stop[i + 1] = stop[i + 2] = false;
        kabu -= 3;
    }
  }
  cout << damage << endl;
}
